game-core/src/main/java/games/strategy/engine/framework/startup/mc/ServerModel.java

```diff
@@ -483,27 +477,19 @@ public IServerMessenger getMessenger() {
   }
 
   public Map<String, String> getPlayersToNodeListing() {
-    synchronized (this) {
-      return new HashMap<>(playersToNodeListing);
-    }
+    return Collections.unmodifiableMap(playersToNodeListing);
```

Comment:
Since you are no longer making a defensive copy, what happens if the caller is iterating the collection and then, in a different thread, the collection is modified?  Won't that trigger a `ConcurrentModificationException`?

Note that I'm assuming instances of `ServerModel` can be accessed concurrently given all the trouble someone went to making it (apparently) thread-safe.  If that's not true, we should probably remove all the synchronization since synchronizing in some places and not in others simply makes the code confusing and hard to reason about.

Reply:
I asked myself the same question and I wasn't sure about this either.
The thing is that for most of the methods callers don't even iterate over the collection and just perform a contains check or similar, so an exception is not that likely to occur in the first place, but this doesn't prove anything.
What all callers have in common though is that they all process the collection immediately and don't keep a reference to it somewhere, so the odds of hitting such an exception are low.
But yeah, simply testing if 'it works' doesn't guarantee anything.
I'm sure the whole synchronisation thing was just over-cautiously added here, but I'll try to analyse the call stack to see if we're safe here (maybe the list is guaranteed to be populated before the methods can be called), but feel free to prove me wrong.

Reply:
I suspect the existing synchronization logic was incorrect, example:
```
(1) Thread A executes the previous code and exits the synchronized block
(2) Thread B concurrently modifies `playersToNodeListing`
(3) The return value from Thread A is now incorrect.
```

It seems like for this to be correct the caller needs to hold the sync lock, we seem to release it too soon.

Otherwise, IMO the vast majority of TripleA locking/sychronization could and should be removed. But, network synchronization is difficult and causes bad errors that are hard to find.

Reply:
I'm wondering how this PR/discussion thread moves forward. 

I think we can guarantee we will not get a `ConcurrentModificationException`
if we can verify that all consumers of this map do not modify it. If so, any such consumers could be updated to make a copy and modify the copy.   Thoughts?

Reply:
Well I verified all consumers calling this method don't actually modify it 🤷‍♂️

Reply:
> I think we can guarantee we will not get a `ConcurrentModificationException`
> if we can verify that all consumers of this map do not modify it.

That's incorrect.  A CME is thrown if, while one thread is iterating a collection, another thread modifies it.  It doesn't require both threads to modify the collection.  From the Javadocs:

> _For example, it is not generally permissible for one thread to modify a Collection while another thread is iterating over it._

<hr>

> The thing is that for most of the methods callers don't even iterate over the collection and just perform a contains check or similar

A "contains check" is typically implemented via iteration.

<hr>

Without a lot of research, we have to assume that instances of this class are accessed concurrently based on its current implementation.  Therefore, I'd just leave the defensive copies in place to be safe.

Reply:
Is everyone ok when I just wrap the ArrayList in a synchronised wrapper for now?

Reply:
1fb45c900c82a5958dcb84a069b0f91ddcd1fa46

Reply:
@RoiEXLab IMO rolling back the synchronization removal would be easiest path forward. Lock/synchronization removal is probably better off in its own PR and individually as each needs consideration :disappointed: